lib/lxdclient.js

```diff
@@ -0,0 +1,428 @@
+        bestFile = items['lxd_combined.tar.gz'];
+        maniFile = bestFile;
+        maniFile.fingerprint = bestFile.sha256;
+
+    } else {
+        maniFile = items['lxd.tar.xz'];
+        if (!maniFile) {
+            throw new restifyErrors.InvalidContentError(
+                'No "lxd.tar.gz" item found for image %s', img.fullname);
+        }
+
+        bestFile = items['squashfs'] || items['root.tar.xz'] ||
+            items['disk-kvm.img'] || items['uefi1.img'] || items['disk1.img'];
+
+        if (!bestFile) {
+            throw new restifyErrors.InvalidContentError(
+                'No filesystem found for image %s', img.fullname);
+        }
+
+        if (bestFile.ftype === 'root.tar.xz') {
+            if (maniFile.lxdhashsha256rootxz) {
+                maniFile.fingerprint = maniFile.combined_rootxz_sha256;
+            } else {
+                maniFile.fingerprint = maniFile.combined_sha256;
+            }
+        } else if (bestFile.ftype == 'squashfs') {
+            maniFile.fingerprint = maniFile.combined_squashfs_sha256;
+        } else if (bestFile.ftype == 'disk-kvm.img') {
+            maniFile.fingerprint = maniFile['combined_disk-kvm-img_sha256'];
+        } else if (bestFile.ftype == 'disk1.img') {
+            maniFile.fingerprint = maniFile['combined_disk1-img_sha256'];
+        } else if (bestFile.ftype == 'uefi1.img') {
+            maniFile.fingerprint = maniFile['combined_uefi1-img_sha256'];
+        } else {
+            throw new restifyErrors.InternalError(
+                'No fingerprint for image "%s" with ftype "%s"',
+                img.fullname, bestFile.ftype);
+        }
+    }
+
+    console.log(JSON.stringify(bestFile, null, 2));
+
+    assert.object(bestFile, 'bestFile');
+    assert.object(maniFile, 'maniFile');
+    assert.string(maniFile.fingerprint, 'maniFile.fingerprint');
+
+    // Cache it.
+    img.bestFile = bestFile;
+    img.manifestFile = maniFile;
+
+    return bestFile;
+}
+
+LxdClient.prototype.getFileStreamForPath =
+function getFileStreamForPath(filePath, callback) {
+    assert.string(filePath, 'filePath');
+
+    this._makeHttpRequest(filePath, function _onHttpResponse(err, req, res) {
+        if (res && !err) {
+            res.pause();
+        }
+        callback(err, res);
+    });
+};
+
+//---- exports
+
+module.exports = {
+    DEFAULT_REGISTRIES: DEFAULT_REGISTRIES,
+    LxdClient: LxdClient,
+    isCombinedImage: isCombinedImage
+};
+
+//---- cmdline
+
+if (require.main === module) {
+    var bunyan = require('bunyan');
+    var fs = require('fs');
+    var tmp = require('tmp');
+
+    var log = bunyan.createLogger({
+        level: (process.env['LOG_LEVEL'] || 'debug'),
+        name: 'lxd images'
+    });
+    var lxdOpts = {
+        log: log,
+        url: 'https://us.images.linuxcontainers.org'
+        // url: 'https://cloud-images.ubuntu.com/releases'
+    };
+    var wantedImage = 'alpine/3.12';
+    // var wantedImage = 'ubuntu:focal:amd64:default';
```

Comment:
These commented-out ones are for the future, right?

Reply:
...I actually have no idea what that's about. Presumably that was just Todd doing testing? It seems to work with arbitrary images as-is, but if that's the case then we should be able to just remove that stuff?

Reply:
This is for command line (standalone) testing - note that for JS if you see this:
```
if (require.main === module) {
...
}
```
then it's being invoked manually as `node lib/lxdclient.js`.